portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -20,89 +20,116 @@
 import com.google.appengine.api.datastore.DatastoreServiceFactory;
 import com.google.appengine.api.datastore.Entity;
 import com.google.appengine.api.datastore.PreparedQuery;
 import com.google.appengine.api.datastore.Query;
 import com.google.appengine.api.datastore.Query.SortDirection;
+import java.time.format.DateTimeFormatter;
+import java.time.LocalDateTime; 
+import java.lang.Math;
 
-/** Servlet that handles comment data. */
+/** 
+  * Servlet that uploads and retrieves persistent comment data using datastore.
+  */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
+  /** Represents a single comment */
+  private static class Comment {
+      String text;
+      String name;
+      String time;
+
+      Comment(String text, String name, String time) {
+        this.text = text;
+        this.name = name;
+        this.time = time;
+      }
+  }
+
+  /** Extracts user comment from form and stores it via datastore. */
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
     String userComment = request.getParameter("text-input");
-    if (userComment != null) {
-      storeComment(userComment);
+    String userName = request.getParameter("author");
+    String timestamp = Long.toString(System.currentTimeMillis());
+
+    if (userComment != null || userComment.trim() == "") {
+      storeComment(userComment, userName, timestamp);
     }
     response.sendRedirect("/index.html");
   }
 
-  /**
-    * Stores user comment as entity in datastore.
-    */
-  private void storeComment(String userComment) {
+  /** Stores user comment as entity in datastore. */
+  private void storeComment(String userComment, String userName, 
+    String timestamp) {
     Entity commentEntity = new Entity("Comment");
     commentEntity.setProperty("text", userComment);
+    commentEntity.setProperty("name", userName);
+    commentEntity.setProperty("time", timestamp);
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     datastore.put(commentEntity);
   }
 
+  /** 
+    * Loads all user comments from datastore and returns JSON list of n 
+    * comments, where n is the number of comments the user has requested. 
+    */
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
     PreparedQuery results = getAllComments();
     
-    ArrayList<String> comments = new ArrayList<String>();
+    ArrayList<Comment> comments = new ArrayList<Comment>();
     int numberToDisplay = getNumberToDisplay(request); 
     int numberDisplayed = 0;
 
-    for(Entity comment : results.asIterable()) {
-      String commentText = (String) comment.getProperty("text");
-      comments.add(commentText);
-      numberDisplayed += 1;
-      if (numberDisplayed == numberToDisplay) {
-        break;
-      }
-    }
+    results.asIterable().forEach(comment -> {
+      comments.add(new Comment(
```

Comment:
with a fromEntity() static factory method, mentioned above, this could be results.asIterable().forEach(comment -> Comment.fromEntity(comment));

Reply:
Changed!